pkg/utils/nodeutils.go

```diff
@@ -115,6 +121,80 @@ func IsMcfgPoolUsingKC(pool *mcfgv1.MachineConfigPool) (bool, string, error) {
 	return true, currentKCMC, nil
 }
 
+func AreKubeletConfigsRendered(pool *mcfgv1.MachineConfigPool, client runtimeclient.Client) (bool, error) {
+	// find out if pool is using a custom kubelet config
+	isUsingKC, currentKCMCName, err := IsMcfgPoolUsingKC(pool)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if pool %s is using a custom kubelet config: %w", pool.Name, err)
+	}
+	if !isUsingKC || currentKCMCName == "" {
+		return true, nil
+	}
+	// if the pool is using a custom kubelet config, check if the kubelet config is rendered
+	kcmcfg := &mcfgv1.MachineConfig{}
+	err = client.Get(context.TODO(), types.NamespacedName{Name: currentKCMCName, Namespace: pool.Namespace}, kcmcfg)
+	if err != nil {
+		return false, fmt.Errorf("failed to get machine config %s: %w", currentKCMCName, err)
+	}
+
+	kc, err := GetKCFromMC(kcmcfg, client)
+	if err != nil {
+		return false, fmt.Errorf("failed to get kubelet config from machine config %s: %w", currentKCMCName, err)
+	}
+
+	var obj interface{}
+	err = json.Unmarshal(kcmcfg.Spec.Config.Raw, &obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to unmarshal machine config %s: %w", currentKCMCName, err)
+	}
+
+	encodedKC, err := jsonpath.Get("storage.files[0].contents.source", obj)
+	if err != nil {
+		return false, fmt.Errorf("failed to get encoded kubeletconfig %s: %w", currentKCMCName, err)
+	}
+	encodedKCStr := encodedKC.(string)
+	if encodedKCStr == "" {
+		return false, fmt.Errorf("failed to get encoded kubeletconfig %s: %w", currentKCMCName, err)
+	}
+	mcPayloadPrefix := `data:text/plain,`
+	encodedKCStr = strings.TrimPrefix(encodedKCStr, mcPayloadPrefix)
+
+	decodedKC, err := url.PathUnescape(encodedKCStr)
+	if err != nil {
+		return false, fmt.Errorf("failed to decode kubeletconfig %s: %w", currentKCMCName, err)
+	}
+
+	fmt.Println(decodedKC)
+	isSubset, diff, err := JSONIsSubset(kc.Spec.KubeletConfig.Raw, []byte(decodedKC))
+	if err != nil {
+		return false, fmt.Errorf("failed to check if kubeletconfig %s is subset of %s: %w", currentKCMCName, decodedKC, err)
+	}
+	if isSubset {
+		return true, nil
+	} else {
+		PrintJSONDiff(diff)
+		return false, fmt.Errorf("kubeletconfig %s is not subset of %s", currentKCMCName, decodedKC)
+	}
+}
+
+func GetKCFromMC(mc *mcfgv1.MachineConfig, client runtimeclient.Client) (*mcfgv1.KubeletConfig, error) {
+	if mc == nil {
+		return nil, fmt.Errorf("machine config is nil")
+	}
+	if len(mc.GetOwnerReferences()) != 0 {
+		if mc.GetOwnerReferences()[0].Kind == "KubeletConfig" {
```

Comment:
Of this list of references, we only check the first? Or does this mean it's safe to assume there is only ever one element in the list?

Reply:
It is safe to assume there is only ever one element in the list as only one KC corresponds to one MC